pyoozie/tags.py

```diff
@@ -504,26 +505,25 @@ class _AbstractWorkflowEntity(typing.Iterable):
 
     def __init__(
             self,
-            xml_tag,       # typing.Text
+            xml_tag=None,  # type: typing.Optional[typing.Text]
             name=None,     # type: typing.Optional[typing.Text]
             on_error=None  # type: typing.Optional[_AbstractWorkflowEntity]
     ):
         # type: (...) -> None
         self.__xml_tag = xml_tag
+        self.__name = name if name else uuid.uuid4().hex[:8]
         self.__on_error = copy.deepcopy(on_error)
-        if name:
-            self.__identifier = '{tag}-{name}'.format(tag=xml_tag, name=name)
-        else:
-            self.__identifier = '{tag}-{uid}'.format(tag=xml_tag, uid=uuid.uuid4().hex[:8])
-
-    def identifier(self):
-        # type: () -> typing.Text
-        return self.__identifier
 
     def xml_tag(self):
         # type: () -> typing.Text
         return self.__xml_tag
 
+    def identifier(self, xml_tag=None):
+        # type: (typing.Optional[typing.Text]) -> typing.Text
+        if not xml_tag:
+            xml_tag = self.__xml_tag
+        return validate_xml_id('{tag}-{name}'.format(tag=xml_tag, name=self.__name))
```

Comment:
Refactored because for the "parallel" collection we need to actually create two identifiers, but we want them to share the same "name", e.g. `fork-0001` and `join-0001` are two different identifiers for two different tags, but we want to be able to relate that these tags are linked together (by sharing the name `0001`).